### DISCUSSION TITLE
Timeout unexpectedly not raised

### DISCUSSION THREAD

**Discussion by valsteen** [encode/httpx#2861] [Potential Issue] [answered]
Hi!

I was investigating why I sometimes have requests that take much longer than the timeout. I could isolate and reproduce the scenario:

I have those versions, on Python 3.11.3:

```
httpx 0.25.0 The next generation HTTP client.
├── certifi *
├── httpcore >=0.18.0,<0.19.0
│   ├── anyio >=3.0,<5.0 
│   │   ├── idna >=2.8 
│   │   └── sniffio >=1.1 
│   ├── certifi * 
│   ├── h11 >=0.13,<0.15 
│   └── sniffio ==1.* (circular dependency aborted here)
├── idna *
└── sniffio *
```

let's have this minimal server which just sleeps 2 seconds

```python
import asyncio
from fastapi import FastAPI

app = FastAPI()


@app.get("/")
async def root():
    await asyncio.sleep(2)
```

I name it `simple.py` and run `uvicorn simple:app`

And here is the test:

```python
import asyncio
import time

from httpx import Limits, AsyncClient, Timeout


async def read():
    start = time.time()
    try:
        response = await http.get("http://127.0.0.1:8000")
        response.raise_for_status()
    except Exception as ex:
        print(f"failed: {ex.__class__}")

    print(f"time: {time.time()-start}")


async def main():
    tasks = []
    for x in range(100):
        tasks.append(asyncio.create_task(read()))
    await asyncio.gather(*tasks, return_exceptions=True)


if __name__ == "__main__":
    limits = Limits(
        max_keepalive_connections=None,
        max_connections=3,
    )
    http = AsyncClient(
        limits=limits,
        timeout=Timeout(5, read=5, pool=5),
    )

    asyncio.get_event_loop().run_until_complete(main())
```

I get timings way above 5 seconds, but I don't understand why ; I would expect a pool timeout.

I can probably work around this by having a timeout on the task itself ( like here: `asyncio.wait_for(read(), 5)` ), but did I misunderstand the functionality, or is this scenario not supported ?

**Comment by zanieb** [comment]
You don't get a timeout because the server is not blocked when sleeping, the connection is free for reuse and the server is processing many new requests concurrently. If you change to a blocking sleep e.g. `time.sleep(2)` you'll start encountering `ReadTimeout` and `PoolTimeout` client-side.

I believe each timeout is enforced individually rather than cumulatively over the entire lifecycle of the request. I'm not sure if there's a great way to limit to the total request time — although it seems like there should be.

**Reply by valsteen** [reply]
Thank you,

I have this situation: 
- a backend endpoint makes requests to a 3rd party. those requests are done through a shared httpx client.
- the 3rd party can be slow, and as a result pending outbound requests accumulate
- if I have many incoming requests pending this can fill up memory ; I'd expect the timeout to mitigate this, so incoming requests that can't be served timely will return early instead of piling up.

Now indeed the timeout is not a sufficient guarantee if I have a large number of inbound connections, but I wouldn't easily hit a memory limit if the pool timeout would raise after the given timeout, no matter how the 3rd party is implemented.

In practice, it's as if individual `http.get` have their pool timeout reset every time one connection from the pool is freed, regardless whether the freed connection will be used by that particular `http.get`.

I would tend to believe httpx is often used in the above scenario, with similar expectations about pooling and timeouts, and this is probably worth documenting.

What can also be confusing for library users, is that `aiohttp` with a similar configuration will actually behave according to the above expectation ; in the below example, a `ServerTimeoutError` will be raised in all tasks that take more than 5 seconds:

```python
import asyncio
import time

from aiohttp import ClientSession, ClientTimeout, TCPConnector


async def read(client: ClientSession):
    start = time.time()
    try:
        response = await client.get("http://127.0.0.1:8000")
        response.raise_for_status()
    except BaseException as ex:
        print(f"failed: {ex.__class__}")

    print(f"time: {time.time()-start}")


async def main():
    async with ClientSession(
        timeout=ClientTimeout(connect=5, sock_read=5, sock_connect=5),
        connector=TCPConnector(limit=3),
    ) as client:
        tasks = []
        for x in range(100):
            tasks.append(asyncio.create_task(read(client)))
        await asyncio.gather(*tasks, return_exceptions=True)


if __name__ == "__main__":
    asyncio.get_event_loop().run_until_complete(main())
```

**Reply by valsteen** [reply]
The possibly misleading definition:

> The pool timeout specifies the maximum duration to wait for acquiring a connection from the connection pool. If HTTPX is unable to acquire a connection within this time frame, a PoolTimeout exception is raised. 

https://www.python-httpx.org/advanced/#fine-tuning-the-configuration

**Comment by valsteen** [comment]
For reference, here is a possible implementation for library users that would like to timeout based on the total request time:

```python
import asyncio
import time
from asyncio import wait_for

from httpx import (
    Limits,
    AsyncClient,
    Request,
    Response,
    AsyncHTTPTransport,
    TimeoutException,
)


class TimingOutTransport(AsyncHTTPTransport):
    def __init__(self, timeout: float, /, *args, **kwargs) -> None:
        self.timeout = timeout
        super().__init__(*args, **kwargs)

    async def handle_async_request(self, request: Request) -> Response:
        try:
            return await wait_for(super().handle_async_request(request), self.timeout)
        except TimeoutError as exc:
            await self.aclose()
            raise TimeoutException(str(exc), request=request) from exc


async def read():
    start = time.time()
    try:
        response = await http.get("http://127.0.0.1:8000")
        response.raise_for_status()
    except Exception as ex:
        print(f"failed: {ex.__class__}")

    print(f"time: {time.time()-start}")


async def main():
    tasks = []
    for x in range(100):
        tasks.append(asyncio.create_task(read()))
    await asyncio.gather(*tasks, return_exceptions=True)


if __name__ == "__main__":
    limits = Limits(
        max_keepalive_connections=None,
        max_connections=3,
    )
    http = AsyncClient(
        transport=TimingOutTransport(5, limits=limits),
    )

    asyncio.get_event_loop().run_until_complete(main())
```

edits: 
- AsyncClient instantiation was missing 
- added `aclose` to free the connection, because cancelling won't free the connection otherwise: https://github.com/encode/httpx/issues/1451#issuecomment-763110954

**Comment by valsteen** [comment]
I believe I found it, this is in fact located in httpcore.

The `while True` in the snippet below will unblock when a connection is freed, but it can still fail with `ConnectionNotAvailable`. It will then again `status.wait_for_connection` for the same timeout value, not accounting for the time that was already spent

https://github.com/encode/httpcore/blob/94ffb33cf378e41662c90615458db9e42b43cad5/httpcore/_async/connection_pool.py#L229-L235

I posted this finding in httpcore: https://github.com/encode/httpcore/discussions/814

**Comment by valsteen** [comment, marked as answer]
this is solved:

- issue: https://github.com/encode/httpcore/issues/822
- PR : https://github.com/encode/httpcore/pull/823
